src/freertos_drivers/esp_idf/EspIdfWiFi.cxx

```diff
@@ -0,0 +1,1355 @@
+            nvs_commit(cfg_);
+            break;
+        default:
+            LOG(LEVEL_ERROR, "wifi: Error %s getting userCfg_.",
+                esp_err_to_name(result));
+            break;
+    }
+
+    // Private configuration.
+    len = sizeof(privCfg_);
+    result = nvs_get_blob(cfg_, NVS_KEY_LAST_NAME, &privCfg_, &len);
+    switch (result)
+    {
+        case ESP_OK:
+            if (len == sizeof(privCfg_))
+            {
+                // Already initialized.
+                break;
+            }
+            // Not initialized yet.
+            // fall through
+        case ESP_ERR_NVS_NOT_FOUND:
+            // fall through
+        case ESP_ERR_NVS_INVALID_LENGTH:
+            // Reset to factory defaults.
+            memset(&privCfg_, 0, sizeof(privCfg_));
+            nvs_set_blob(cfg_, NVS_KEY_LAST_NAME, &userCfg_, sizeof(userCfg_));
+            nvs_commit(cfg_);
+            break;
+        default:
+            LOG(LEVEL_ERROR, "wifi: Error %s getting privCfg_.",
+                esp_err_to_name(result));
+            break;
+    }
+}
+
+//
+// EspIdfWiFi::init_wifi()
+//
+void EspIdfWiFi::init_wifi(WlanRole role)
+{
+    if (role == WlanRole::DEFAULT_ROLE)
+    {
+        role = userCfg_.mode_;
+    }
+
+    ESP_ERROR_CHECK(esp_netif_init());
+
+    /// @todo Do we need this?
+    esp_err_t result = esp_event_loop_create_default();
+    if (result != ESP_OK)
+    {
+        LOG(LEVEL_ERROR, "wifi: esp_event_loop_create_default() failed: %s",
+            esp_err_to_name(result));
+    }
+
+    ESP_ERROR_CHECK(esp_event_handler_instance_register(
+        WIFI_EVENT, ESP_EVENT_ANY_ID, &wifi_event_handler, this, nullptr));
+    ESP_ERROR_CHECK(esp_event_handler_instance_register(
+        IP_EVENT, ESP_EVENT_ANY_ID, &ip_event_handler, this, nullptr));
+
+    // Use this to debug wifi connection problems
+    // esp_log_level_set("wifi", ESP_LOG_VERBOSE);
+
+    wifi_init_config_t cfg = WIFI_INIT_CONFIG_DEFAULT();
+    ESP_ERROR_CHECK(esp_wifi_init(&cfg));
+
+
+    // Set and initialize the appropriate role(s).
+    switch (role)
+    {
+        default:
+            // Fall through.
+        case WlanRole::UNKNOWN:
+            // Fall through.
+        case WlanRole::AP_STA:
+            ESP_ERROR_CHECK(esp_wifi_set_mode(WIFI_MODE_APSTA));
+            init_softap(userCfg_.ap_.ssid_, userCfg_.ap_.pass_);
+            init_sta();
+            break;
+        case WlanRole::AP:
+            ESP_ERROR_CHECK(esp_wifi_set_mode(WIFI_MODE_AP));
+            init_softap(userCfg_.ap_.ssid_, userCfg_.ap_.pass_);
+            break;
+        case WlanRole::STA:
+            ESP_ERROR_CHECK(esp_wifi_set_mode(WIFI_MODE_STA));
+            init_sta();
+            break;
+    }
+    // Start the WiFi.
+    ESP_ERROR_CHECK(esp_wifi_start());
+
+    if (apIface_)
+    {
+        esp_netif_ip_info_t ip_info;
+        esp_netif_get_ip_info(
+            esp_netif_get_handle_from_ifkey("WIFI_AP_DEF"), &ip_info);
```

Comment:
we may want to move the IF names to a constant or better if we can reference their own constants (haven't looked for one though).